File: sdks/typescript/src/Logger.ts
import type { LoggerOptions, Sink } from './types.js';
import { cloneStructured, normalizeEnvironment } from './internal/snapshot.js';
import { WideEvent } from './WideEvent.js';

function resolveService(service?: string): string {
  const resolved = service?.trim() || process.env.OPENAKTA_SERVICE?.trim();
  if (!resolved) {
    throw new Error('Logger requires a service name or OPENAKTA_SERVICE');
  }

  return resolved;
}

function resolveEnvironment(environment?: string): string {
  return normalizeEnvironment(environment ?? process.env.OPENAKTA_ENV);
}

export class Logger {
  private readonly service: string;
  private readonly environment: string;
  private readonly sinks: readonly Sink[];
  private readonly defaultContext: Record<string, unknown>;

  constructor(options: LoggerOptions) {
    this.service = resolveService(options.service);
    this.environment = resolveEnvironment(options.environment);
    this.sinks = [...(options.sinks ?? [])];
    this.defaultContext = cloneStructured(options.defaultContext ?? {});
  }

  /** Create a new WideEvent lifecycle for one operation. */
  startEvent(operation: string): WideEvent {
    const event = new WideEvent(operation, this.service, this.environment, [...this.sinks]);
    if (Object.keys(this.defaultContext).length > 0) {
      event.appendContext(this.defaultContext);
    }
    return event;
  }

  /** Wrap an async or sync function in a WideEvent lifecycle automatically. */
  async trace<T>(
    operation: string,
    fn: (event: WideEvent) => Promise<T> | T,
  ): Promise<T> {
    const event = this.startEvent(operation);

    try {
      const result = await fn(event);
      await event.emit({ status: 'ok' });
      return result;
    } catch (err) {
      event.setError(err);
      await event.emit();
      throw err;
    }
  }
}

File: sdks/csharp/Openakta.Logger/Logger.cs
using Openakta.Logger.Sinks;

namespace Openakta.Logger;

public sealed class Logger
{
    private readonly string _service;
    private readonly string _environment;
    private readonly IReadOnlyList<ISink> _sinks;
    private readonly Dictionary<string, object?> _defaultContext;

    public Logger(LoggerOptions options)
    {
        _service = ResolveService(options.Service);
        _environment = ResolveEnvironment(options.Environment);
        _sinks = options.Sinks?.ToList() ?? [];
        _defaultContext = new Dictionary<string, object?>(options.DefaultContext ?? new Dictionary<string, object?>());
    }

    internal Logger(string service, string environment, IReadOnlyList<ISink> sinks, IReadOnlyDictionary<string, object?> defaultContext)
    {
        _service = service;
        _environment = environment;
        _sinks = sinks;
        _defaultContext = new Dictionary<string, object?>(defaultContext);
    }

    public static LoggerBuilder Builder() => new();

    public WideEvent StartEvent(string operation)
    {
        var eventInstance = new WideEvent(operation, _service, _environment, _sinks);
        if (_defaultContext.Count > 0)
        {
            eventInstance.AppendContext(_defaultContext);
        }

        return eventInstance;
    }

    public async Task<T> TraceAsync<T>(string operation, Func<WideEvent, Task<T>> fn)
    {
        var eventInstance = StartEvent(operation);
        try
        {
            var result = await fn(eventInstance).ConfigureAwait(false);
            await eventInstance.EmitAsync(status: "ok").ConfigureAwait(false);
            return result;
        }
        catch (Exception ex)
        {
            eventInstance.SetError(ex);
            await eventInstance.EmitAsync().ConfigureAwait(false);
            throw;
        }
    }

    private static string ResolveService(string? candidate)
    {
        var resolved = !string.IsNullOrWhiteSpace(candidate) ? candidate : Environment.GetEnvironmentVariable("OPENAKTA_SERVICE");
        if (string.IsNullOrWhiteSpace(resolved))
        {
            throw new InvalidOperationException("Logger requires a service name or OPENAKTA_SERVICE");
        }

        return resolved;
    }

    private static string ResolveEnvironment(string? candidate)
    {
        var resolved = !string.IsNullOrWhiteSpace(candidate) ? candidate : Environment.GetEnvironmentVariable("OPENAKTA_ENV");
        return resolved is "production" or "staging" or "development" ? resolved : "production";
    }
}
